### PR TITLE
0.2 Major Refactoring

### PR DESCRIPTION
**Pull request by cscherrer**
# `FixedRNG`

`FixedRNG()` is an `AbstractRNG` that is, well, fixed. It always returns the same value. Like this:
```julia
julia> rand(FixedRNG())
0.5

julia> randn(FixedRNG())
0.0

julia> randexp(FixedRNG())
1.0
```

One possibility for this is to move it to MeasureBase (code is short), and use it to define a fallback method for `testvalue`.

# `rand`
- now disallowed on `ModelPosterior`
- `rand` returns a sample from the latent space. To get a return value, call `predict`
 
# Return-related functions
- `drop_return` is now `dropreturn`
- `hasreturn(m)` returns either `HasReturn()` or `NoReturn()`, depending whether the model contains a `return` statement

# Configurations (`<:AbstractConfig`)

I've saved the best for last. There's kind of a lot here, so let's just walk through a simple example, `logdensity_def`.

First we define 
```julia
struct LogdensityConfig{F} <: AbstractConfig
    f::F
end
```
`AbstractConfig` subtypes are used for dispatch in Tilde, and can also hold information to be referenced as the model is run.

Next, here's the top-level call:
```julia
function MeasureBase.logdensityof(
    cm::AbstractConditionalModel,
    pars::NamedTuple;
)
    cfg = LogdensityConfig(logdensityof)
    runmodel(cfg, cm, pars, (ℓ=0,))
end
```

`AbstractConditionalModel` can be either `ModelClosure` (a model with specified argument) or `ModelPosterior` (if it has also been conditioned on some observations).

The `logdensityof` call is very simple: it just sets the config and then calls `runmodel`, which does the work. `runmodel` takes four arguments:
1. The configuration
2. An `AbstractConditionalModel`
3. `pars`, which is just another place to specify values of more variables. This is more efficient than pulling the model apart and rebuilding it each time with different values. We expect `pars` to change often.
4. A "context", which we usually shorten to `ctx`. This is typically a NamedTuple. It's different from a config in that a context is expected to be updated at each `~` statement in the model.

Now, if a model we're using happens to have a `return` statement, we need to be sure not to call it. To make this more general, we replace every `return r` with `return retfun(cfg, r, ctx)`. If the model has no `return` statement, a model will implicitly return a named tuple of all variables appearing on the left of a `~`.

Say our model has an observed value `x`, and contains the statement
```julia
x[j] ~ Normal()
```

Tilde first rewrites this (using `Tilde.opticize`) as
```julia
(x, @optic _[j]) ~ Normal()
```
Next, we often want different behavior depending whether a value has been observed. For this we have an abstract type `MaybeObserved`, with instances `Observed` and `Unobserved`. 

Since `x` is observed, we represent it as `Observed{:x}(x)`. This is a handy way to pass the information
1. This variable has been observed
2. Its name is `:x`
3. Its value is `x`

Finally, the line `x[j] ~ Normal()` is converted to

```julia
(x, ctx) = tilde(cfg, Observed{:x}(x), (@optic _[j]), Normal(), ctx)
```

The `tilde` function needs to dispatch on the type of `cfg` (and possibly other values), and return an updated `x` and context.

In this case, that function looks like this:
```julia
@inline function tilde(
    cfg::LogdensityConfig{typeof(logdensityof)},
    obj::MaybeObserved{X},
    lens,
    d,
    ctx::NamedTuple,
) where {X}
    x = value(obj)
    pred = predict(d, lens(x))
    @reset ctx.ℓ += logdensityof(d, lens(x))
    (pred, ctx)
end
```

That `pred` line is a little weird. `d` might itself be a model, in which case our observations would be of the latent variables of that model. For the case `d` is _not_ a model, there's a fallback method, `predict(m, x) = x`.

### CODE REVIEW

File: src/Tilde.jl [hunk 1]
```diff
@@ -79,29 +82,33 @@ include("callify.jl")
...
 include("primitives/testvalue.jl")
-include("primitives/testparams.jl")
-include("primitives/weightedsampling.jl")
+# include("primitives/testparams.jl")
+# include("primitives/weightedsampling.jl")
 include("primitives/measures.jl")
-include("primitives/basemeasure.jl")
+# include("primitives/basemeasure.jl")
+include("primitives/predict.jl")
+# include("primitives/dag.jl")
```
Comment (jamesthesnake) [inline src/Tilde.jl:110]: Question, why not just delete this as oppose to comment it out ?

Comment (cscherrer) [reply src/Tilde.jl:110]: Oh, I didn't know anyone was watching this!

I'm not permanently removing that, just temporarily while I get other things to work. It's a fairly big refactoring

File: src/fixedrng.jl [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+Random.randexp(::FixedRNG, ::Type{T}) where {T<:Real} = one(T)
+
+# We need concrete type parameters to avoid amiguity for these cases
+for T in [Float16, Float32, Float64]
+    @eval begin
+        Base.rand(::FixedRNG, ::Type{$T}) = one($T) / 2
+        Random.randn(::FixedRNG, ::Type{$T}) = zero($T)
+        Random.randexp(::FixedRNG, ::Type{$T}) = one($T)
+    end
+end
```
Comment (cscherrer) [inline src/fixedrng.jl:19]: @oschulz What do you think about moving this to MeasureBase? Then we'd have an easy fallback `testvalue` implementation:
```julia
testvalue(m) = rand(FixedRNG(), m)
```

File: src/optics.jl [hunk 3]
```diff
@@ -43,6 +43,17 @@ end
...
+#     setindex!(o, val, l.pure.indices...)
+# end
+
+# # Attempting to set a value outside the current eltype widens the eltype
+# @inline function _setindex!(o::AbstractArray{T}, val::V, l::Lens!!{<:IndexLens}) where {T,V}
+#     new_o = similar(o, Union{T,V})
+#     new_o .= o
+#     setindex!(new_o, val, l.pure.indices...)
+# end
+
```
Comment (cscherrer) [inline src/optics.jl:56]: Can we get rid of this?

File: src/primitives/basemeasure.jl [hunk 4]
```diff
@@ -1,43 +1,44 @@
...
+) where {X}
+    x = value(obj)
+    (x, ctx)
 end
+
+function rproduct(nt::NamedTuple)
+    productmeasure(map(rproduct, nt))
+end
+
+rproduct(m::AbstractMeasure) = m
```
Comment (cscherrer) [inline src/primitives/basemeasure.jl:44]:
```suggestion
rproduct(m::AbstractMeasure) = m

```

File: benchmarks/bouncy.jl [hunk 5]
```diff
@@ -14,56 +14,73 @@ using ForwardDiff
...
+        u.value, u.partials[]
+    end
+
+    gconfig = ForwardDiff.GradientConfig(obj, rand(d), ForwardDiff.Chunk{d}())
+    function ∇neglogp!(y, t, x, args...)
+        ForwardDiff.gradient!(y, obj, x, gconfig)
+        y
+    end
+    ℓ, dneglogp, ∇neglogp!
+end
```
Comment (cscherrer) [inline benchmarks/bouncy.jl:40]: I think this was just reorganization with no substantive changes, but we need to make sure

File: src/GG/deprecated_codes/static_closure_conv.jl [hunk 6]
```diff
@@ -39,6 +39,11 @@ function mk_closure_static(expr, toplevel::Vector{Expr})
                                         $Closure{$glob_name,typeof(frees)}(frees)
                                     end
                                 )
+                                ret = :(
+                                    let frees = $closure_arg
+                                        $Closure{$glob_name,typeof(frees)}(frees)
+                                    end
+                                )
```
Comment (cscherrer) [inline src/GG/deprecated_codes/static_closure_conv.jl:46]: Need to review this

File: src/core/utils.jl [hunk 7]
```diff
@@ -156,35 +152,47 @@ end
...
+    #         push!(loader.args, :($k = Tilde.NestedTuples.lazymerge(_obs.$k, _pars.$k)))
+    #     else
+    #         T = getproperty(pars, k)
+    #         push!(loader.args, quote
+    #             _k = $qk
+    #             @warn "Duplicate key, ignoring $_k in data"
+    #             $k::$T = _pars.$k
+    #         end)
+    #     end
+    # end
```
Comment (cscherrer) [inline src/core/utils.jl:181]: Do we need type declarations for observations and parameters?

File: src/lensvars.jl [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+            _ => begin
+                foreach(f, args)
+            end
+        end
+
+        return result
+    end 
+    
+    foldast(leaf, branch)(opticize(ast))
+end
```
Comment (cscherrer) [inline src/lensvars.jl:24]:
```suggestion
end

```

File: test/runtests.jl [hunk 9]
```diff
@@ -168,34 +163,29 @@ include("examples-list.jl")
...
+        end
+    
+        @test rand(m()).x isa Bool
+        @test logdensityof(m(), rand(m())) isa Float64
     end
 
-    @test rand(m(); ctx=()) isa Bool
-    @test logdensity(m(), rand(m())) isa Float64
-end
+end
```
Comment (cscherrer) [inline test/runtests.jl:191]:
```suggestion
end

```

File: src/Tilde.jl [hunk 10]
```diff
@@ -79,29 +79,33 @@ include("callify.jl")
...
-# include("primitives/basemeasure.jl")
 include("primitives/testvalue.jl")
-include("primitives/testparams.jl")
-include("primitives/weightedsampling.jl")
+# include("primitives/testparams.jl")
+# include("primitives/weightedsampling.jl")
 include("primitives/measures.jl")
 include("primitives/basemeasure.jl")
+include("primitives/predict.jl")
+# include("primitives/dag.jl")
```
Comment (cscherrer) [inline src/Tilde.jl:107]: Still need to update some methods

File: benchmarks/bouncy.jl (hunk 5 above)
Comment (cscherrer) [reply benchmarks/bouncy.jl]: Dropping these changes, they really should be separate

**Pull request merged**